plugins/node/opentelemetry-instrumentation-awslambda/src/awslambda.ts

```diff
@@ -0,0 +1,178 @@
+import { VERSION } from './version';
+
+export class AwsLambdaInstrumentation extends InstrumentationBase {
+  constructor() {
+    super('@opentelemetry/instrumentation-awslambda', VERSION);
+  }
+
+  init() {
+    // _HANDLER and LAMBDA_TASK_ROOT are always defined in Lambda.
+    const taskRoot = process.env.LAMBDA_TASK_ROOT!;
+    const handlerDef = process.env._HANDLER!;
+
+    const handler = path.basename(handlerDef);
+    const moduleRoot = handlerDef.substr(0, handlerDef.length - handler.length);
+
+    const [module, functionName] = handler.split('.', 2);
+
+    // Lambda loads user function using an absolute path.
+    let filename = path.resolve(taskRoot, moduleRoot, module);
+    if (!filename.endsWith('.js')) {
+      // Patching infrastructure currently requires a filename when requiring with an absolute path.
+      filename += '.js';
+    }
+
+    return [
+      new InstrumentationNodeModuleDefinition(
+        // NB: The patching infrastructure seems to match names backwards, this must be the filename, while
+        // InstrumentationNodeModuleFile must be the module name.
+        filename,
+        ['*'],
+        undefined,
+        undefined,
+        [
+          new InstrumentationNodeModuleFile(
+            module,
+            ['*'],
+            (moduleExports: LambdaModule) => {
+              diag.debug('Applying patch for lambdatest handler');
+              if (isWrapped(moduleExports[functionName])) {
+                this._unwrap(moduleExports, functionName);
+              }
+              this._wrap(moduleExports, functionName, this._getHandler());
+              return moduleExports;
+            },
+            (moduleExports?: LambdaModule) => {
+              if (moduleExports == undefined) return;
+              diag.debug('Removing patch for lambdatest handler');
+              this._unwrap(moduleExports, functionName);
+            }
+          ),
+        ]
+      ),
+    ];
+  }
+
+  private _getHandler() {
+    return (original: Handler) => {
+      return this._getPatchHandler(original);
+    };
+  }
+
+  private _getPatchHandler(original: Handler) {
+    diag.debug('patch handler function');
+    const plugin = this;
+
+    return function patchedHandler(
+      this: {},
+      event: {},
+      context: Context,
+      callback: Callback
+    ) {
+      const name = context.functionName;
+      const span = plugin.tracer.startSpan(name, {
+        kind: SpanKind.SERVER,
+        attributes: {
+          [FaasAttribute.FAAS_EXECUTION]: context.awsRequestId,
+          [FaasAttribute.FAAS_ID]: context.invokedFunctionArn,
+          // TODO(anuraaga): Add cloud.account.id when there is a JS accessor for it.
+        },
+      });
+
+      // Lambda seems to pass a callback even if handler is of Promise form, so we wrap all the time before calling
+      // the handler and see if the result is a Promise or not. In such a case, the callback is usually ignored.
```

Comment:
> In such a case, the callback is usually ignored.

What if the callback isn't ignored? Would the span be ended twice/would that matter?